### DISCUSSION TITLE
Azure Deployment

### DISCUSSION THREAD

**Discussion by aderici** [Budibase/budibase#5886] [Show and tell]
Hi,

Starting this thread to help Azure people like me to deploy a simple self hosted Budibase into Azure App Services. 

App services is the platform as a service offering of MS that is the go-to service if you simply want to run your app and not deal with anything else. It supports running codes directly (like a node.js app from github) or it can run a single docker container as well as simple docker-compose files (with a lot of limitations which makes running budibase impossible)

Another interesting limitation is in appservice if we want to use a custom docker image wa have to write our important data into /home folder otherwise azure wont persist any other files. 

If we want to use direct ssh into docker feature we have to install openssh and provide a static user pass (app runs behind reverse proxy and ssh port is never exposed to public)

Given all info and limitations, i pursued the single docker path. Thankfully the single docker folder under hosting in the repo was a helpful start but definitely not working out of the box.  After a many sleepless nights i was able to get it to work. 

One thing i couldnt understand yet is if i build and run the app and worker from the repo, app doesnt work and complain "global db not found" i had to grab official app docker and copy the already compiled app directory into my single docker. Any idea why that might be happening?

2 of my clients have been using single docker budibase on azure happily ever since. Upgrades will be painful though. 

I'd love to help single docker efforts to provide something usable for the community, please guide me how,

**Comment by mike12345567** [comment]
Hi @aderici - the single docker image is on our roadmap for official support soon, right now it was just put together as an experiment to see how it would work and what it might look like, but isn't officially released yet!

In terms of the build failing, I'm not sure what branch of Budibase you're using to build with, but `master` is the branch you should be using, there may be issues with the `develop` branch as it is still in development - we have it as the default branch so that external contributions will be added to it so we can test them before release.

As I say the single docker image is something that we want to support soon, we're still working out the details of documenting it, everything we want to expose via environment variables and putting together some testing for it, once this is done we'll have it released.

**Comment by PekkaJalonen** [comment]
@aderici I'm also looking an option to run on appservice the single container, can you give some high level insights on what type of customization did you need for the build to get it working?

**Comment by aderici** [comment]
@PekkaJalonen i pretty much had to go through every file on hosting/single folder on this repo and make changes. Nginx config, dockerfile, the text indexing plugin config file, startup script and others. 2 main issues i had to solve, one if them is persisting files into /home folder and other is just getting it to work. Also installed openssh to support app service ssh. I hope some day we will have a standard single docker supported by budibase which also fulfills azure appservice requirements,

**Comment by jonnymccullagh** [comment]
Hi @aderici, we have added a step in our build process that creates a single image with some Azure App Service customizations. It is tagged as budibase-aas if you want to try it out and give us some feedback:
```
docker pull registry.hub.docker.com/budibase/budibase-aas
```

**Reply by aderici** [reply]
Hey thats great to hear! Really happy to see continuous improvement here! I'll try it out and will let you know

**Reply by PekkaJalonen** [reply]
It would be nice to get some configuration info also. I have set minio, couch, redis, jwt, internal api key related env variables and tested mounting /data to file storage and also without mount using the native persistent storage, and I get nginx error 502 Bad Gateway, so no luck. Tested with v1.2.2

**Reply by PekkaJalonen** [reply]
Also seems that SSH is not working, so either it has configuration issue or the container is down. Last log couchdb cluster stable.

**Comment by PekkaJalonen** [comment]
Tested the budibase-aas image on Azure App Service. Could not get the container working when deployed using Single container, only when using Docker Compose to deploy single container which is in preview in Azure.

From functionality perspective, App Service SSH is not connecting and also, minio works only when not using persistent storage. Each time trying to mount /data to Azure File Storage or using the App Service native storage(home folder), I start getting errors from minio about file access denied. I can see couch, minio and .env generated, but minio not storing anything.

**Comment by jonnymccullagh** [comment]
Hi guys, I have merged pull requests to help support running Budibase in Azure App Service. This included the SSH service on port 2222, /home for data and some manipulation of environment variables as AAS already set a variable named PORT. 
Install instructions are here:
https://docs.budibase.com/docs/azure-app-service
Note the docker image address of `budibase/budibase-aas` 
We would appreciate any testing or feedback you might have.

@aderici @PekkaJalonen

**Reply by aderici** [reply]
Hi, @jonnymccullagh just started using it for one of the projects a few days ago. all seems to be working just fine other than SSH. See #7745. I just checked docker hub and the latest image seems from 2 days ago so i assume your merge didnt automatically generate new images in docker hub?

**Reply by jonnymccullagh** [reply]
Good point. Triggered it now.

**Reply by aderici** [reply]
@jonnymccullagh just switched to the latest and seems to be working fine including the ssh! Thank you!

**Reply by PekkaJalonen** [reply]
budibase-aas release works fine when running on Azure App Service in non-persistent mode, but when enabling data persistence by setting WEBSITES_ENABLE_APP_SERVICE_STORAGE=true, minio fails to store files due to fact that App Service used file storage does not comply well with applications which require file handles/locks as described in  https://learn.microsoft.com/en-us/azure/app-service/configure-connect-to-azure-storage?pivots=container-linux&tabs=portal "It's not recommended to use storage mounts for local databases (such as SQLite) or for any other applications and components that rely on file handles and locks."

But when running on example Azure Container Apps and mounting file storage, minio works fine. So most likely this is App Service limitation which you cannot fix as minio has no configuration of disabling such locks. I recommend using Container Apps or changing budibase to use additional options to replace minio with example https://github.com/gaul/s3proxy which would allow defining different S3 targets including Azure Blob Storage.

**Comment by PekkaJalonen** [comment]
Hi, I started another discussion around the Azure before I realized that this was here already, so I'm just linking my finding here: #8040

**Reply by aderici** [reply]
I have been usingg budibase with azure appservice happily and havent noticed issues with file storage. Are you sure its incompatible? I can add attachments and retrieve them back successfully

**Reply by PekkaJalonen** [reply]
Have you enabled the file storage persistence which is by default set to false on app service. I kow budibase works as expected when it is off, but also it does not persist your data is it is off. You can restart the site while it is off, and your data will be available, but in case there is bigger change to the site, tou will loose you data. See https://learn.microsoft.com/en-us/azure/app-service/configure-custom-container?pivots=container-linux#use-persistent-shared-storage

**Reply by aderici** [reply]
Yes i always set it to true but the doc also states it is by default "on" for linux custom containers which is what we are using right?

By default, persistent storage is enabled on Linux custom containers. To disable it, set the WEBSITES_ENABLE_APP_SERVICE_STORAGE app setting value to false via the Cloud Shell.

Are there specific steps i can follow to reproduce an error case?

**Reply by PekkaJalonen** [reply]
I tested deploying to different sites and it was set false for me, I don't know why. And have you checked from the kudu files, that minio folders are actually there? Se my post 7 days ago to reproduce the issue https://github.com/Budibase/budibase/discussions/5886#discussioncomment-3707853 When setting the app service storage true, files are written to /home. These are visible under kudu bash file browsing. Minio fails when it tries to write in app service file storage.

**Reply by jonnymccullagh** [reply]
For me the default was WEBSITES_ENABLE_APP_SERVICE_STORAGE was false. Maybe I could have ignored that but when I tried setting it to true the container restarted and I had problems. I was then hoping that I could use an Azure File Storage service to mount a persistent file system to /home but I found that mounting to /home wasn't possible. It may have been possible with /home/budibase but I have to get back to that as I was on holiday for a week.

**Reply by PekkaJalonen** [reply]
I have tested all possible combinations, unfortunately I cannot see that it will work with App Service, /home is the path when using app service file storage. You can also mount separate file storage to app service container and I tested that as well using /data path with the normal budibase/budibase image. In both cases, the Minio issue happens. So it is not limitation of Azure File Storage as it works with AKS and ACA, but App Service limitation on how the mount is done.

**Reply by aderici** [reply]
![image](https://user-images.githubusercontent.com/6830693/193083217-b4eca0a6-f2f7-41ec-8054-7c25fd97acf3.jpeg)

I can confirm i have a client on appservice with the storage env variable set to true. I have checked the other issue number you mentioned. There are no steps to reproduce the error. Can you tell me exact steps to take so that i can see an error message? I can add attachments to my records without problems.

**Reply by PekkaJalonen** [reply]
Hi, I thought that I might made some misconfigurations in my environments, so I tested by following the Budibase docs exactly by creating new resource groups and everything from scratch as described, and then when signing in, tested uploading new logo and it failed. I used westeurope region for all resources.

**Reply by PekkaJalonen** [reply]
I created issue of the problem #7482

**Reply by aderici** [reply]
Mount volume ${WEBAPP_STORAGE_HOME}:/data


Can you tell me about this? It makes me feel you are using docker compose on appservice instead of custom container

I dont specify any paths and by default the budibase standard image will store its files under /home which will be persisted

**Reply by PekkaJalonen** [reply]
I tested with both, compose(using WEBAPP_STORAGE_HOME) and single docker using basic app service storage /home + additionally testing with single docker using the "bring your own storage" mounting /data. For /data mount I used the budibase/budibase version and when testing /home path i used budibase/budibase-aas image. All resulted with same issue.

**Reply by PekkaJalonen** [reply]
@aderici did you deploy using cli or ARM etc, or using portal? I used portal to deploy as per following the docs in budibase docs exactly. I'm just thinking that might the different deployment type cause the persistence issue as it is not that rare that portal and cli deployments provides small differences in deployments.

**Reply by PekkaJalonen** [reply]
And the portal method sets the persistence to false same as for jonnymccullagh who mentioned same behavior. Also I tried testing different App Service Environment sizes to see is that would have impact, but nothing solved the issue. So I would appreciate if you could give input how you made the deployment, so I could try replicate functioning deployment.

**Reply by PekkaJalonen** [reply]
Interesting note which I found at https://github.com/hashicorp/terraform-provider-azurerm/issues/15014

This is why I mentioned I use west europe:

Important Factoids

Running in West Europe.
When creating the Function App through the Azure Web UI, WEBSITES_ENABLE_APP_SERVICE_STORAGE=false is set by default.

**Reply by aderici** [reply]
I used azure portal. I always use us-east. The environment variable of persistent storage was not there when i created the app. I added it manually, setting it to true.

**Reply by PekkaJalonen** [reply]
Ok, i will do some testing tomorrow by deploying to westeurope and some other region to see if there is differences. But I can verify, that I have always used westeurope and the WEBSITES_ENABLE_APP_SERVICE_STORAGE is always there, but it has default value false.

**Reply by PekkaJalonen** [reply]
Ok, so made some testing. I deployed new app service to eastus, francecentral and westeurope following exactly those steps in budibase docs and leaving other thing as default. For all three sites the WEBSITES_ENABLE_APP_SERVICE_STORAGE was there, but set to false. For each site I changed it to true. After signin, went to organization settings and tried to upload a new logo > Budibase gives error and throws me out.

**Reply by jonnymccullagh** [reply]
I've added a PR here to try to force the WEBSITES_ENABLE_APP_SERVICE_STORAGE variable to true before the container service runs it in `build-target-paths.sh` and also doubled-down by adding it to `runner.sh` too. Hoping this gets us a little further if you guys want to comment on the PR:
https://github.com/Budibase/budibase/pull/8059

**Reply by PekkaJalonen** [reply]
From my point of view, setting this WEBSITES_ENABLE_APP_SERVICE_STORAGE to true is not an big issue and can be fixed by adding the step to instructions, but point is that when it is set to true, Budibase will fail as Minio does not work with the storage. So far I have only heard from @aderici that he has successfully created App Service docker which works with persistent storage. For me all deployments fails. Additionally I have read about other Minio deployments to App Service which do not work. So until there is instructions how to deploy it successfully with persistent storage, is does not work in App Service.

**Reply by aderici** [reply]
@PekkaJalonen i'm really curious about how mine is working just fine. I believe the error message you mentioned was about file / folder permissions access right? How did you corelate this with azure app service storage and file locks? Could it be a folder ownership/permission issue? I believe minio is not only used for attachments but also for the actual app assets like js css etc. So if all else is working just fine but not the attachments could it be some other cause?

**Reply by PekkaJalonen** [reply]
Hi, the issue shows file access permission error. Trying to upload logo or create a new app from template fails, so anything trying to write files to storage. @aderici I would appreciate if you could test this deployment with budibase instructions and give feedback if it work for you or not. Also, I'm curious to know which version of budibase you have deployed and also, if you open app service "Advanced Tools" > Go > Bash > open directory /home and check the file structure (there should be the minio folders) to see that files are actually written there and buckets exists?!

**Reply by aderici** [reply]
@PekkaJalonen i opened app service panel and used "ssh" option to dive into the container. The "couch" folder is underneath /home. I can see all my attachments under folders like this :  /prod-budi-app-assets/app_xxx/attachments/xxx.png

Permissions of the couch folder is like this : 

```drwxrwxrwx   2 nobody nogroup        0 Sep 13 00:36 couch```

I have also checked the folders from advanced tools - bash, same results. Also i did type "printenv" into this bash console and saw that "APPSETTING_WEBSITES_ENABLE_APP_SERVICE_STORAGE=true"

The bash is not docker shell its a different environment that just has access to the same storage. Creating / modifying files from bash can cause file permission issues.

I see that my environment variable for persistent storage is set but how can i be sure if its persistent? I know for a fact that i have upgraded my system (changed docker image) multiple times and i didnt lose data.

Here's my current version and how its set up : 

<img width="553" alt="image" src="https://user-images.githubusercontent.com/6830693/193632843-685a4b6c-92ea-4218-afcd-e7e570b22555.png">

If i can find some time i'll set up one from scratch to see if i face any issues

**Reply by PekkaJalonen** [reply]
I think that checking the app service storage /home content via FTP or from the advanced tools\bash is the best way to see that files actually exists in storage. When using ssh to budibase itself you can find files in /home, but does not guarantee that those are in the app service storage.

**Reply by aderici** [reply]
@PekkaJalonen i did check both, files are there. One more thing, how about app service tier? I'm using production tier, which sometimes has differences to a dev/test instance : 

<img width="1618" alt="image" src="https://user-images.githubusercontent.com/6830693/193636907-71c2b940-4e16-473e-9051-05492d5af908.png">

**Reply by PekkaJalonen** [reply]
Ok, so it could work then. I tested that as well, B1 and also with P1V2

**Reply by PekkaJalonen** [reply]
@aderici just to check, you mention couch folder and permissions, but the issue is with minio. Can you check minio folder permission?

**Reply by jonnymccullagh** [reply]
@aderici my Azure UI looks very different from your screenshot above. I recorded a video of what I did here: https://www.youtube.com/watch?v=wgOU6lgeUvU (apologies for the quality). 
When I ran the container in the above video I was able to upload a PDF file and viewed it on the filesystem via SSH. However, after restarting the container all the content was gone. I am interested in how @aderici is using AAS if differently from me. And interested in how @PekkaJalonen is getting minio errors if he is following the same process as in the video.

**Reply by aderici** [reply]
@jonnymccullagh the UI is actually the same. my screenshot is taken after deploying the web app, browsing to the deployment center on the left hand menu. One thing i noticed is that the owner of the folders look different. 

@PekkaJalonen sorry i said couch but i meant minio. 

Here's what my ls-al on prod budi app assets looks like : 
<img width="544" alt="image" src="https://user-images.githubusercontent.com/6830693/193666191-fcf21056-e843-4723-a71e-620d7aaf905a.png">

I'll record a loom video deploying an appservice from scratch today

**Reply by PekkaJalonen** [reply]
@jonnymccullagh did you set the WEBSITES_ENABLE_APP_SERVICE_STORAGE to true on App Service configuration? My step were the same with addition to set this mentioned setting to true which was false by default. And I also noticed you used ssh to connect to budibase. This does not tell you if the /home folder is persistent. I can do the exact same steps as you did as the service uses temporary volumes when the previously mentioned setting is false. This is why you also loose your data. When setting it to true, /home is mounted in App Service used storage which will persist you data. Here is when the minio gets broken for me.

**Reply by jonnymccullagh** [reply]
No I didn't set WEBSITES_ENABLE_APP_SERVICE_STORAGE to true. At what point are you setting that? And I will try the same.

**Comment by aderici** [comment]
@jonnymccullagh @PekkaJalonen after some research i have good and bad news. 

It seems that minio recently introduced some changes, which determines how the files are stored in the disk. Previously if you have a single server your files would be stored as is, meaning if you upload a.jpg you'd see an a.jpg in your disk. But now, if you save a.jpg you see a folder called a.jpg in your disk and within that folder there are files that specify the content. 

This behavior is determined by the content of /path-to-minio-data/.minio.sys/format.json. See how in an earlier successful deployment, "format" property is "fs". 

```{"version":"1","format":"fs","id":".....","fs":{"version":"2"}}```

If you have previously deployed a budibase and have been upgrading, minio will see your existing folder / format.json and will continue with the fs format and that will work fine with azure. But it wont allow you to start a new instance with fs format anymore

See how the new format.json looks on a recent deployment, format is now "xl-single"

```
{"version":"1","format":"xl-single","id":".....,"xl":{"version":"3","this":"....","sets":[["....."]],"distributionAlgo":"SIPMOD+PARITY"}}
```

Minio won't allow you to manually change a new deployments format, it will throw an error. It will only allow you to migrate an old folder with fs format. It is their way of fighting agaist people using fs mode i guess

I have been building my custom images earlier, when we didnt have budibase-aas standard image. When i checked still have an old image that works out of the box and the minio version on that one is "RELEASE.2022-05-23T18-45-11Z"

@jonnymccullagh would it be possible to use this specific version of minio i mentioned in azure images?

Are you guys using minio for budibase cloud? Or are you using s3 and if we somehow have an s3 compatible proxy for azure, can we ditch minio completely by setting some environment variables?

**Reply by PekkaJalonen** [reply]
Thank you for solving the root cause. This seems to be limitation with App Service then, because I can use Azure Container App successfully with Minio when mounting to Azure File Storage share.

I also tested the older version with Minio gateway to blob storage which works well with Azure Blob Storage in App Service. Bad news is that Minio has dropped the support and futher development of the gateways, so it is not sustainable solution.

**Reply by PekkaJalonen** [reply]
@aderici I started think about the permissions of the folders you mentioned earlier that your ls -al looks different and the fact that Minio works with ACA (so it cannot be Azure File Storage limitation). Could it be also that Minio changed permissions somehow for the new deployments. I think that we could try to test permission fix and see if that helps. It could be then fixed in budibase if that might be the issue.

**Reply by aderici** [reply]
Unfortunately this seems to be a filesystem issue as you previously highlighted. I've tried chmod 777 or chown but no luck.

**Reply by PekkaJalonen** [reply]
Yes, App Service mounts with owner nobody / nogroup which cannot be changed. In ACA mount is owner root / root which seems to be the reason why it works there.

**Reply by aderici** [reply]
@PekkaJalonen there is a discussion about file locking which is very similar to the error messages that i see. But this is an old message and we have been using it happily on fs mode. https://github.com/minio/minio/issues/9212

**Comment by jonnymccullagh** [comment]
@PekkaJalonen @aderici I have been playing with Azure Container Instances as an alternative to AAS. I created a video here: https://www.youtube.com/watch?v=40EzlgftPo0
Not sure that this totally ticks all the boxes for you but it does appear to persist the storage and minio uploads working fine. 
I'll try to also get a look over ACA as I get time.

**Reply by aderici** [reply]
Thank you @jonnymccullagh nice demonstration, happy to see at least another path exists. @PekkaJalonen i see that mounting a storage into appservice is possible (into a custom local folder, not /home) and when doing so, you can either mount a blob storage or file storage. Shouldnt "file storage" work fine with appservice and minio as it is exactly the same type of storage that you use for container instances? note that if you mount a "blob" type of storage its not going to work because blob storages are read only when mounted to appservice

**Reply by PekkaJalonen** [reply]
Unfortunately this does not help. I tested it with single container as well compose and having some custom folder path with budibase/budibase-aas release as well with the budibase/budibase mounting to /data. Both gives the same issue. App Service still mounts with user/group = nobody/nogroup as where when using ACA the mount is root/root which I assume is the reason for the problem.

**Reply by PresidentCamach0** [reply]
Hi, is there any more updates to this thread?

We want to deploy on Azure Web Apps.
When we set WEBSITES_ENABLE_APP_SERVICE_STORAGE to TRUE, BudiBase works for about a day or two. Once the service restarts, it keeps the logins and apps. When you try to open an app it shows the error: Error initialising app: Not Found 

Is this still an issue with Azure Web Apps or the version of CouchDB being incompatible?

**Comment by PresidentCamach0** [comment]
Hi, is there any more updates to this thread?

We want to deploy on Azure Web Apps.
When we set WEBSITES_ENABLE_APP_SERVICE_STORAGE to TRUE, BudiBase works for about a day or two. Once the service restarts, it keeps the logins and apps. When you try to open an app it shows the error: Error initialising app: Not Found

Is this still an issue with Azure Web Apps or the version of CouchDB being incompatible?

**Reply by mike12345567** [reply]
Hi @OriginalLimited - this sounds like a problem with the MinIO storage being wiped out after the service restarts. We encountered issues with this when using Azure Web Apps, sadly we were unable to find a good solution to this.

**Reply by jonnymccullagh** [reply]
Azure offer approx 3 container services each with different pros & cons. I think the best option is [Azure Container Instances](https://docs.budibase.com/docs/azure-container-instances) for a persistent storage volume.

**Reply by PresidentCamach0** [reply]
@jonnymccullagh Thank you, and thanks for your YouTube guides on setting this up. The only thing with the container instances is the additional work behind setting up SSL and custom domains - which, out of the box on Azure Web Apps is very simple with a couple of clicks.

In the end, we just spun up an Ubuntu VM and installed Budibase on there - works much better and easier to setup SSL/Custom Domains.

**Reply by PresidentCamach0** [reply]
@mike12345567 Thanks! That's a shame!

**Reply by aderici** [reply]
I remember being able to run budibase on azure appservice. Its been more than a year though. Just be careful where individual services save their files. When you enable persistent storage only home folder is persisted. So you have to make sure all important files are saved underneath /home/ path

**Comment by PresidentCamach0** [comment]
Re-visiting this thread, it appears recently that the budibase/budibase-aas container has been updated for MinIO for it to run using persistent storage using Azure Blob.

I can confirm it works, however the setup information is not correct on the Budibase setup guides here: https://docs.budibase.com/docs/azure-app-service

Currently, it only works using the experimental tag, and at version 3.4.13. Using stable does not work.

But otherwise, so far so good. It loads fine even after the app service is restarted/stopped.

**Comment by ff-lkubin** [comment]
Hi. Fairly new to this platform so I apologize if I am missing something simple here. Deploying via Azure App Services as well, the application deploys but once inside budibase deploying / importing apps returns "Unable to write to object store bucket." I am misinterpreting something in the instructions here?

**Reply by PresidentCamach0** [reply]
The deployment instructions on the BudiBase website do not work. Currently, it only works using the experimental tag, and at version 3.4.13.

I've had bad test results using Azure Web Apps.
It might keep the data persistently when the web service restarts, but I often find that the BB server does not start and you get a bunch of errors in the log stream. I would only use it in development and not production.

![image](https://github.com/user-attachments/assets/c6bf51d8-9dc2-4b3a-9874-5cce0ceeea38)
![image](https://github.com/user-attachments/assets/e11c9a35-945a-4833-86f1-c66d762cc9b6)
